Alter additional properties during the transition (#9)
im-open/transition-jira-tasks-by-query: README.md

# transition-jira-tasks-by-query

This GitHub Action will query Jira using JQL provided as an input, and will transition any tickets it finds to a given status.

## Requirements

> This action requires access to our on-prem network for accesss to the Jira servers.  You must run the action with a `self-hosted` or `coldstone` runner.

Reviewer: Since this is a public action we should avoid talking about our setup directly. Instead maybe we can make this more generic and call out that if the Jira server is hosted on an internal network then the action must run on a runner that has access to the network.

As an aside, we are trying to stop using the term `coldstone`. I'm blanking on the new term, but using `self-hosted` is good enough.

Author: Good point.  Will make the adjustment.

| `fail-if-issues-not-found` | false          | The action will fail if no issues are found. Otherwise, the action will succeed if no issues are found. _`false` by default._                                                                                                    |
| `fail-jira-inaccessible`   | false          | Fail if Jira is inaccessible at the moment. Sometimes Jira is done but shouldn't block the pipeline. _`false` by default._                                                                                                       |
| `jira-username`            | false          | The username to login to Jira with in order to perform the transition. _Will be ignored if not set._                                                                                                                             |
| `jira-password`            | false          | The password to login to Jira with in order to perform the transition. _Must be set if `jira-username` is set. If set when `jira-username` is not set, it will be ignored._                                                      |

> <sup>*</sup> Either `jql-query` or `issues` input is required.  If both are provider, `jql-query` will be used.

Reviewer: ❤️

  - [License](#license)

## Inputs

| Parameter                          | Is Required    | Description                                                                                                                                                                                                                      |
|------------------------------------|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `domain-name`                      | true           | The domain name for Jira.                                                                                                                                                                                                        |
| `jql-query`                        | conditionally* | The JQL query to use to find tickets that will be transitioned. A max of 20 issues can be transitioned.                                                                                                                          |

> | `issues`                           | conditionally* | Comma delimited list of issues to transition. Use `im-open/get-workitems-action` to identify list of issues for a PR or deployment.                                                                                              |

Author: The project uses issues, tickets and tasks  terms to reference a Jira Issue.

Would it be possible to consolidate on one term?  Notice the repo name as well is using task as a term.

Work Item is what the `im-open/get-work-items-action` uses.  But it is more generic.

Reviewer: That's a good point. This action was intended to transition deployment tasks which is why it has the name it does. Maybe that's not the best term to use any longer though. It seems like the Jira documentation leans towards issues, so maybe that's the way to go.

Author: You ok if I update the naming to _issues_ in another PR?  Or would you rather I do it in this PR?  Renaming the repo should be backwards compatible.  Something you can do at any time.

Reviewer: We can save the rename for another PR. I think that would be better.

| `domain-name`                      | true           | The domain name for Jira.                                                                                                                                                                                                        |
| `jql-query`                        | conditionally* | The JQL query to use to find tickets that will be transitioned. A max of 20 issues can be transitioned.                                                                                                                          |
| `issues`                           | conditionally* | Comma delimited list of issues to transition. Use `im-open/get-workitems-action` to identify list of issues for a PR or deployment.                                                                                              |
| `transition-name`                  | true           | The name of the transition to perform. _Examples might include Open, In Progress, Deployed, etc._                                                                                                                                |

> | `update-fields`                    | false          | A [map](#updating-fields) of issue screen fields to overwrite, specifying the sub-field to update and its static value(s) for each field. When multiple sub-fields or other operations are required, use 'update' input instead. |
> | `process-operations`               | false          | A [map](#updating-fields) containing the field name and a list of operations to perform. _The fields included in here cannot be included in 'fields' input._                                                                     |

Reviewer: Is this suggestion right? I think in some of the naming changes these two input names didn't get updated.

```suggestion
| `update-fields`                    | false          | A [map](#updating-fields) of issue screen fields to overwrite, specifying the sub-field to update and its static value(s) for each field. When multiple sub-fields or other operations are required, use 'process-operations' input instead. |
| `process-operations`               | false          | A [map](#updating-fields) containing the field name and a list of operations to perform. _The fields included in here cannot be included in 'update-fields' input._                                                                     |
```

Author: You are right. That was a miss.

> <sup>*</sup> Either `jql-query` or `issues` input is required.  If both are provider, both will be included.

## Outputs

| Output               | Description                                                                                | Type                 |
|----------------------|--------------------------------------------------------------------------------------------|----------------------|
| `processed-issues`   | Issues successfully transitioned, skipped and (if enabled) with an unavailable transition. | Comma-delimited list |
| `failed-issues`      | Issues in Jira not successfully processed.                                                 | Comma-delimited list |
| `unavailable-issues` | Issues missing the specificed transition.                                                  | Comma-delimited list |
| `not-found-issues`   | Issues not found in Jira.                                                                  | Comma-delimited list |

> | `is-successful`      | One or more issues were transitioned successfully.                                         | Boolean              |

Reviewer: It's also successful if all issues are skipped right?

Author: Yes.

| Parameter         | Is Required | Description                                                                                                                                                           |
| ----------------- | ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `domain-name`     | true        | The domain name for Jira.                                                                                                                                             |
| `jql-query`       | maybe       | The JQL query to use to find tickets that will be transitioned. If provided, will override any issues passed in the `issues` input.                                   |

> | `issues`          | maybe       | Comma delimited list of issues to transition. Use `im-open/get-workitems-action` to identify list of issues for a PR or deployment                                    |

Reviewer: I like the idea of indicating in the "Is Required" column that one of these needs to be specified. I think in our other actions we just put `false` and add something to the description about it.

I think instead of `maybe` something like `conditionally*` might work better. And then preface the note at the bottom with a `*`. I think using that well known pattern would be helpful. Numerical superscript would also work, but we only have one item here so 🤷

| `transition-name` | true        | The name of the transition to perform. Examples might include Open, In Progress, Deployed, etc.                                                                       |
| `jira-username`   | false       | The username to login to Jira with in order to perform the transition. Will be ignored if not set.                                                                    |
| `jira-password`   | false       | The password to login to Jira with in order to perform the transition. Must be set if jira-username is set. If set when jira-username is not set, it will be ignored. |

> Either `jql-query` or `issues` input must be provided.

## Outputs

| Output     | Description            |
|------------|------------------------|

> | `identified-issues-from-query` | Issues from query or issues that exist in Jira from the generated query. |

Reviewer: This is a little easier to understand to me.

```suggestion
| `identified-issues-from-query` | Issues found in Jira using the `jql-query` or `issues` list. |
```

          # issues: |
          #   TW-1234
          #   TW-4455
          # etc.

          # If you want to update or overwrite fields, you can use the following inputs:
          # update-fields: |
          #   {
          #     "customfield_12345": "some value"
          #   } 

      # Some issue types don't have a transition status like others.  
      # In those cases, where you want to still transition, a second step will need to be invoked.  
      # You can pass in the `unavailable-issues` output to transition those remaining issues.
      - name: Transition Remaining Jira Tickets to Closed Status
        uses: im-open/transition-jira-tasks-by-query@v2.0.0
        with:
          jira-username: ${{ vars.JIRA_USERNAME }}
          jira-password: ${{ secrets.JIRA_PASSWORD }}
          domain-name: jira.com
          transition-name: Closed
          issues: ${{ steps.transition.outputs.unavailable-issues }}
```

## Updating Fields

They are two different ways to update fields.  Passing a `update-fields` input or an `process-operations` input. You may not specify the same field name in both the `update-fields` and `process-operations` inputs.

> If you are unsure what field names to use, run the [Get-Jira-Issue](../main/Test-Get-Jira-Issue.ps1) script locally from your machine using your ExtendHealth username and password.

Reviewer: Again, let's avoid using our terminology. Everyone should be familiar with logging into Jira and which credentials are required anyway.

Author: Good point.  Will make the adjustment.
